registry/storage/driver/s3-aws/s3.go

```diff
@@ -498,6 +535,59 @@ func New(params DriverParameters) (*Driver, error) {
 	}, nil
 }
 
+// logS3OperationHandlerName is used to identify the handler used to log S3 API
+// requests
+const logS3OperationHandlerName = "docker.storage-driver.s3.operation-logger"
+
+// logS3Operation logs each S3 operation, including request and response info,
+// as it completes
+func (d *driver) logS3Operation(ctx context.Context) request.NamedHandler {
+	return request.NamedHandler{
+		Name: logS3OperationHandlerName,
+		Fn: func(r *request.Request) {
+			req := r.HTTPRequest
+			resp := r.HTTPResponse
+			op := r.Operation
+			fields := map[interface{}]interface{}{
+				"s3_operation_name":                        op.Name,
+				"s3_http_request_method":                   req.Method,
+				"s3_http_request_host":                     req.Host,
+				"s3_http_request_path":                     req.URL.Path,
+				"s3_http_request_content-length":           req.ContentLength,
+				"s3_http_request_remote-addr":              req.RemoteAddr,
+				"s3_http_response_header_x-amz-request-id": resp.Header.Values("x-amz-request-id"),
+				"s3_http_response_status":                  resp.StatusCode,
+				"s3_http_response_content-length":          resp.ContentLength,
+			}
+
+			for logKey, headerKey := range d.LogS3APIResponseHeaders {
+				values := resp.Header.Values(headerKey)
+				if len(values) == 1 {
+					fields[logKey] = values[0]
+					continue
+				}
+				if len(values) > 0 {
+					fields[logKey] = values
+				}
+			}
+
+			ll := dcontext.GetLoggerWithFields(ctx, fields)
+			ll.Info("S3 operation completed")
+		},
+	}
+}
+
+func (d *driver) s3Client(ctx context.Context) *s3.S3 {
```

Comment:
Possible optimization: when `LogS3APIRequests` is false, this function could return `d.S3`, avoiding the copy. The copy is not that expensive, but it's nice to retain the existing copy-free behavior when we can anyway.